src/test/java/org/springframework/cloud/deployer/spi/cloudfoundry/CloudFoundryTaskLauncherIntegrationTests.java

```diff
@@ -0,0 +1,257 @@
+		envProperties.put(CloudFoundryDeployerProperties.MEMORY_PROPERTY_KEY, "1024");
+		envProperties.put(CloudFoundryDeployerProperties.DISK_PROPERTY_KEY, "2048");
+
+		List<String> commandLineArgs = new ArrayList<>(3);
+		commandLineArgs.add("--foo=bar");
+		commandLineArgs.add("--baz=qux");
+		commandLineArgs.add("random=" + UUID.randomUUID().toString());
+
+		request = new AppDeploymentRequest(
+			new AppDefinition("timestamp", Collections.emptyMap()),
+			context.getResource("classpath:batch-job-1.0.0.BUILD-SNAPSHOT.jar"),
+			envProperties,
+			commandLineArgs);
+
+		CloudFoundryOperations cloudFoundryOperations = DefaultCloudFoundryOperations.builder()
+			.cloudFoundryClient(cfAvailable.getResource())
+			.organization("FrameworksAndRuntimes")
+			.space("development")
+			.build();
+
+		taskLauncher = new CloudFoundryTaskLauncher(cfAvailable.getResource(), cloudFoundryOperations, properties);
+	}
+
+	@Test
+	public void testNonExistentAppsStatus() {
+		assertThat(taskLauncher.status("foo").getState(), is(LaunchState.unknown));
+	}
+
+	@Test
+	public void testSimpleLaunch() throws InterruptedException {
+
+		String taskId = taskLauncher.asyncLaunch(request).block(Duration.of(5, ChronoUnit.MINUTES));
+
+		System.out.println(">> taskId = " + taskId);
+
+		TaskStatus status = taskLauncher.asyncStatus(taskId).block();
+
+		while (!status.getState().equals(LaunchState.complete)) {
+			System.out.println(">> state = " + status.getState());
+			Thread.sleep(5000);
+			status = taskLauncher.asyncStatus(taskId).block();
+		}
+
+		assertThat(status.getState(), is(LaunchState.complete));
+	}
+
+	@Test
+	public void testSimpleCancel() throws InterruptedException {
+		Map<String, String> envProperties = new HashMap<>();
+		envProperties.put(CloudFoundryDeployerProperties.SERVICES_PROPERTY_KEY, "my_mysql");
+		envProperties.put(CloudFoundryDeployerProperties.MEMORY_PROPERTY_KEY, "1024");
+		envProperties.put(CloudFoundryDeployerProperties.DISK_PROPERTY_KEY, "2048");
+
+		List<String> commandLineArgs = new ArrayList<>(2);
+		commandLineArgs.add("30000");
+
+		request = new AppDeploymentRequest(
+			new AppDefinition("long-runner", Collections.emptyMap()),
+			context.getResource("classpath:long-running-task-1.0.0.BUILD-SNAPSHOT.jar"),
+			envProperties,
+			commandLineArgs);
+
+		String taskId = taskLauncher.asyncLaunch(request).block(Duration.of(5, ChronoUnit.MINUTES));
+
+		System.out.println(">> taskId = " + taskId);
+
+		Thread.sleep(10000L);
+
+		System.out.println(">> About to cancel the task");
+
+		taskLauncher.cancel(taskId);
+
+		TaskStatus status = taskLauncher.asyncStatus(taskId).block();
+
+		while (!status.getState().equals(LaunchState.failed)) {
+			System.out.println(">> state = " + status.getState());
+			Thread.sleep(5000);
+			status = taskLauncher.asyncStatus(taskId).block();
+		}
+
+		assertThat(status.getState(), is(LaunchState.failed));
+	}
+
+	@After
+	public void cleanUp() throws InterruptedException {
```

Comment:
Unfortunately, I've been using it as a utility method.  When you create a v3 app (task or LRP) and bind it to a service, there is no way to delete that app until you've unbound the services.  Currently the only way to unbind the services is via this code (the CLI doesn't support it and there is no support for v3 apps in any of the CF UI components).  As for each test method doing cleanup, I disagree because if the test fails, the cleanup wouldn't be executed.  Any cleanup belongs in an `@After` method.